### PR TITLE
[ROCm] Added LoadOp and StoreOp implementation for AMDGPU

### PR DESCRIPTION
**Pull request by B1tway**
In this implementation LoadOp and StoreOp are implemented using inline asm

This implementation is temporary because:
* The location and number of s_waincnt is not optimal
* Memory operations are not vectorized 
* Access to memory is non-coalesced

In the future, this solution should be improved and possibly based on the memref or rocdl dialect to allow the compiler to optimize memory accesses without hindrance

### CODE REVIEW

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1034]: `ii` looks ambiguous, can we use `wordIdx` instead?

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1287]: same here. `ii` looks ambiguous, can we use `wordIdx` instead?

**Review by whchung — CHANGES_REQUESTED** [review]
Please supply test cases.

**Comment by whchung** [comment]
@B1tway Please improve the description of the PR. I want to understand: a) how is the implementation be carried out b) what is missing in the current implementation.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp
Comment (whchung) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1031]: @B1tway This hurts performance badly. Can it be removed and let the compiler decides when `s_waitcnt` be inserted?

**Review by whchung — COMMENTED** [review]
Using inline asm to emit load and s_waitcnt is very bad for performance.
Logic here will very likely become throwaway codes.

Even though I can understand the necessity to get something working as a baseline, please help come up with a plan to improve this implementation.

File: lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp
Comment (codego7250) [inline lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:1298]: is this wait_cnt a must?

**Comment by B1tway** [comment]
You can create a kernel.mlir file for testing.
**Command for execution:**
```
python3 triton/tools/aot.py kernel.mlir --target=amdgcn --gfx=908
```
**kernel.mlir:**
```
#blocked0 = #triton_gpu.blocked<{sizePerThread = [2], threadsPerWarp = [64], warpsPerCTA = [2], order = [0]}>
module attributes {"triton_gpu.num-warps" = 2 : i32} {
  func @global_load_store(%arg0: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg1: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg2: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg3: i32) {
    %c256_i32 = arith.constant 256 : i32
    %0 = tt.get_program_id {axis = 0 : i32} : i32
    %1 = arith.muli %0, %c256_i32 : i32
    %2 = tt.make_range {end = 256 : i32, start = 0 : i32} : tensor<256xi32, #blocked0>
    %3 = tt.splat %1 : (i32) -> tensor<256xi32, #blocked0>
    %4 = arith.addi %3, %2 : tensor<256xi32, #blocked0>
    %5 = tt.splat %arg0 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %6 = tt.addptr %5, %4 : tensor<256x!tt.ptr<i8>, #blocked0>
    %7 = tt.splat %arg1 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %8 = tt.addptr %7, %4 : tensor<256x!tt.ptr<i8>, #blocked0>


    %9 = tt.load %6 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xi8, #blocked0>
    %10 = tt.load %8 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xi8, #blocked0>
    %11 = arith.addi %9, %10 : tensor<256xi8, #blocked0>
    %12 = tt.splat %arg2 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %13 = tt.addptr %12, %4 : tensor<256x!tt.ptr<i8>, #blocked0>

    tt.store %13, %11 : tensor<256xi8, #blocked0>
    return
  }
}
```
**Expected output:**
```
        .text
        .amdgcn_target "amdgcn-amd-amdhsa--gfx908:sramecc+:xnack-"
        .globl  global_load_store
        .p2align        8
        .type   global_load_store,@function
global_load_store:
        v_writelane_b32 v8, s6, 0
        s_mov_b64 s[8:9], s[4:5]
        v_readlane_b32 s4, v8, 0
        s_load_dwordx2 s[6:7], s[8:9], 0x0
        s_load_dwordx2 s[10:11], s[8:9], 0x8
        s_load_dwordx2 s[8:9], s[8:9], 0x10
        s_mov_b32 s5, 1
        v_lshlrev_b32_e64 v0, s5, v0
        s_mov_b32 s5, 0xfe
        v_and_b32_e64 v1, v0, s5
        s_mov_b32 s5, 8
        v_mov_b32_e32 v0, s5
        v_lshl_or_b32 v6, s4, v0, v1
        v_ashrrev_i32_e64 v0, 31, v6
        v_mov_b32_e32 v7, v0
        s_waitcnt lgkmcnt(0)
        s_mov_b32 s4, s6
        v_mov_b32_e32 v0, v6
        s_mov_b32 s6, s7
        v_mov_b32_e32 v2, v7
        v_add_co_u32_e64 v0, s[4:5], s4, v0
        v_mov_b32_e32 v1, s6
        v_addc_co_u32_e64 v2, s[4:5], v1, v2, s[4:5]
        v_mov_b32_e32 v1, v2
        ;;#ASMSTART
        global_load_ubyte v5, v[0:1], off offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        s_mov_b32 s4, 8
        v_lshrrev_b16_e64 v4, s4, v5
        s_mov_b32 s6, s10
        v_mov_b32_e32 v0, v6
        s_mov_b32 s5, s11
        v_mov_b32_e32 v2, v7
        v_add_co_u32_e64 v1, s[6:7], s6, v0
        v_mov_b32_e32 v0, s5
        v_addc_co_u32_e64 v0, s[6:7], v0, v2, s[6:7]
        v_mov_b32_e32 v2, v0
        ;;#ASMSTART
        global_load_ubyte v0, v[1:2], off offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        v_lshrrev_b16_e64 v3, s4, v0
        s_mov_b32 s6, s8
        v_mov_b32_e32 v1, v6
        s_mov_b32 s5, s9
        v_mov_b32_e32 v6, v7
        v_add_co_u32_e64 v1, s[6:7], s6, v1
        v_mov_b32_e32 v2, s5
        v_addc_co_u32_e64 v6, s[6:7], v2, v6, s[6:7]
        v_mov_b32_e32 v2, v6
        v_add_u16_e64 v0, v0, v5
        s_mov_b32 s5, 0xff
        v_and_b32_e64 v0, v0, s5
        v_add_u16_e64 v3, v3, v4
        v_lshlrev_b16_e64 v3, s4, v3
        v_or_b32_e64 v0, v0, v3
        s_mov_b32 s4, 0xffff
        v_and_b32_e64 v0, s4, v0
        ;;#ASMSTART
        global_store_byte v[1:2], v0, off offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        s_endpgm
        .section        .rodata,#alloc
        .p2align        6
        .amdhsa_kernel global_load_store
                .amdhsa_group_segment_fixed_size 0
                .amdhsa_private_segment_fixed_size 0
                .amdhsa_kernarg_size 28
                .amdhsa_user_sgpr_count 6
                .amdhsa_user_sgpr_private_segment_buffer 1
                .amdhsa_user_sgpr_dispatch_ptr 0
                .amdhsa_user_sgpr_queue_ptr 0
                .amdhsa_user_sgpr_kernarg_segment_ptr 1
                .amdhsa_user_sgpr_dispatch_id 0
                .amdhsa_user_sgpr_flat_scratch_init 0
                .amdhsa_user_sgpr_private_segment_size 0
                .amdhsa_system_sgpr_private_segment_wavefront_offset 0
                .amdhsa_system_sgpr_workgroup_id_x 1
                .amdhsa_system_sgpr_workgroup_id_y 0
                .amdhsa_system_sgpr_workgroup_id_z 0
                .amdhsa_system_sgpr_workgroup_info 0
                .amdhsa_system_vgpr_workitem_id 0
                .amdhsa_next_free_vgpr 9
                .amdhsa_next_free_sgpr 12
                .amdhsa_reserve_vcc 0
                .amdhsa_reserve_flat_scratch 0
                .amdhsa_reserve_xnack_mask 0
                .amdhsa_float_round_mode_32 0
                .amdhsa_float_round_mode_16_64 0
                .amdhsa_float_denorm_mode_32 3
                .amdhsa_float_denorm_mode_16_64 3
                .amdhsa_dx10_clamp 1
                .amdhsa_ieee_mode 1
                .amdhsa_fp16_overflow 0
                .amdhsa_exception_fp_ieee_invalid_op 0
                .amdhsa_exception_fp_denorm_src 0
                .amdhsa_exception_fp_ieee_div_zero 0
                .amdhsa_exception_fp_ieee_overflow 0
                .amdhsa_exception_fp_ieee_underflow 0
                .amdhsa_exception_fp_ieee_inexact 0
                .amdhsa_exception_int_div_zero 0
        .end_amdhsa_kernel
        .text
.Lfunc_end0:
        .size   global_load_store, .Lfunc_end0-global_load_store

        .section        ".note.GNU-stack"
        .amdgpu_metadata
---
amdhsa.kernels:
  - .args:
      - .address_space:  global
        .offset:         0
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         8
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         16
        .size:           8
        .value_kind:     global_buffer
      - .offset:         24
        .size:           4
        .value_kind:     by_value
    .group_segment_fixed_size: 0
    .kernarg_segment_align: 8
    .kernarg_segment_size: 28
    .max_flat_workgroup_size: 1024
    .name:           global_load_store
    .private_segment_fixed_size: 0
    .sgpr_count:     12
    .sgpr_spill_count: 1
    .symbol:         global_load_store.kd
    .vgpr_count:     9
    .vgpr_spill_count: 0
    .wavefront_size: 64
amdhsa.target:   'amdgcn-amd-amdhsa--gfx908:sramecc+:xnack-'
amdhsa.version:
  - 1
  - 1
...

        .end_amdgpu_metadata
```

**Comment by whchung** [comment]
@B1tway from the ISA dump we can see the undesirable ISA which prevents compiler from doing performance optimization:

```
        ;;#ASMSTART
        global_load_ubyte v5, v[0:1], off offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        ;;#ASMSTART
        global_load_ubyte v0, v[1:2], off offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
```

Had we not been using inline assembly, better sequence could be yielded such as:

```
        global_load_ubyte v5, v[0:1], off offset:0;
        global_load_ubyte v0, v[1:2], off offset:0;
        s_waitcnt vmcnt(0)  ;
```

By not using inline assembly the compiler could do more optimizations moving memory ops around, and potentially vectorize them. And we can save unnecessary `s_waitcnt` instructions which stall the whole program.

For example, with Triton LLVM path, a vector add ISA would be something like:

```
        global_load_dwordx4 v[0:3], v[8:9], off
        global_load_dwordx4 v[4:7], v[8:9], off offset:2048
        global_load_dwordx4 v[8:11], v[18:19], off
        global_load_dwordx4 v[12:15], v[18:19], off offset:2048
        s_waitcnt vmcnt(1)
```

Here we can see the compiler has done several optimizations:
- vectorization using global_load_dwordx4
- reorder VALU instructions so s_waitcnt doesn't have to follow global_load_dwordx4
- vmcnt values of s_waitcnt won't have to be 0 to allow subsequent instructions continue to execute with in-flight memory ops.

None of these can be achieved with the current form of this PR.

One easy way to improve this PR, is to simply use MLIR builder API and emit memref::Load/StoreOp, which would produce regular LLVM load/store instructions, and allow all of the optimizations in the backend. This is how it was done in LLVM path, using LLVM IR builder API.

I can agree to use the current implementation to form a working baseline. But the implementation would surely need to be improved in subsequent PRs.

**Comment by B1tway** [comment]
@whchung Sorry for the wait. 

Thank you for the detailed explanation. I agree with everything.

The inline asm approach for memory instructions causes many optimization problems. Including the problem that the compiler cannot independently arrange the necessary s_waincnt.

The current solution was originally planned as a temporary one. I wrote about it in [closed PR](https://github.com/ROCmSoftwarePlatform/triton/pull/26).

I will update and add the necessary information to the description of this PR in the near future

**Review by whchung — APPROVED** [review]
Approving so we can form a baseline working solution.

**Comment by B1tway** [comment]
@whchung 
I updated the PR description

**Comment by whchung** [comment]
@rsanthanam-amd I'll let you decide whether to take the PR. I'm not familiar with the failures in CI tests.

**Review by rsanthanam-amd — APPROVED** [review]

**Comment by B1tway** [comment]
@rsanthanam-amd 
Tests have been added. Is it possible to merge this PR?

**Review by rsanthanam-amd — APPROVED** [review]

**Pull request merged**
